pkg/providers/amifamily/bootstrap/windows.go

```diff
@@ -44,5 +67,50 @@ func (w Windows) Script() (string, error) {
 		userData.WriteString(fmt.Sprintf(` -ContainerRuntime '%s'`, *w.KubeletConfig.ContainerRuntime))
 	}
 	userData.WriteString("\n</powershell>")
-	return base64.StdEncoding.EncodeToString(userData.Bytes()), nil
+	return userData.String()
+}
+
+// mergeCustomUserData merges XML custom user data with the generated EKS bootstrap script.
+// Merges powershell and script tags, and for all other tags, later arguments in the list take precedence
+func mergeCustomUserData(userDatas ...string) (string, error) {
+	if len(userDatas) == 0 {
+		return "", nil
+	} else if len(userDatas) == 1 {
+		return userDatas[0], nil
+	}
+
+	mergedUserDataXML := &UserDataXML{}
+	for _, userData := range userDatas {
+		// Wrap with userdata tag because encoding/xml package relies on having a well-formed XML with a root element
+		userData = strings.Replace(userData, "&", "&amp;", -1) // Escape ampersands to ensure well-formed XML
+		userData = "<userdata>" + userData + "</userdata>"
+
+		userDataXML := &UserDataXML{}
+		if err := xml.Unmarshal([]byte(userData), &userDataXML); err != nil {
+			return "", err
+		}
+
+		if userDataXML.PowerShell != "" {
+			mergedUserDataXML.PowerShell = userDataXML.PowerShell + mergedUserDataXML.PowerShell
+		}
+		if userDataXML.Script != "" {
+			mergedUserDataXML.Script = userDataXML.Script + mergedUserDataXML.Script
+		}
+		if userDataXML.Detach != "" {
+			mergedUserDataXML.Detach = userDataXML.Detach
+		}
+		if userDataXML.Persist != "" {
+			mergedUserDataXML.Persist = userDataXML.Persist
+		}
+	}
+
+	s, err := xml.Marshal(mergedUserDataXML)
+	if err != nil {
+		return "", err
+	}
+
+	// Remove enclosing <userdata></userdata> tags
+	userdata := strings.TrimPrefix(string(s), "<userdata>")
```

Comment:
Why do we add and then remove these `<userdata>` tags?

Reply:
The reason of wrapping with the userdata string tags is that XML documents [must contain one root element](https://www.w3.org/TR/xml/#:~:text=There%20is%20exactly%20one%20element%2C%20called%20the%20root). As such the `encoding/xml` function `xml.Unmarshal` requires a root element to properly map the XML string data into a struct. The workaround was therefore this rather hacky way you see here.

Other options using a non-standard XML parsing library which is more flexible to parse XML strings without a root.  Another option might be to not map the string into a struct and manipulate the strings directly. Both solutions weren't preferable but happy to discuss options.

Reply:
Removed the flexibility to specify EC2Launch XML elements. In the future if the need arises we can implement further changes to support this.